rustls/src/crypto/aws_lc_rs/hpke.rs

```diff
@@ -0,0 +1,945 @@
+use alloc::boxed::Box;
+use alloc::sync::Arc;
+use alloc::vec::Vec;
+use core::fmt::{self, Debug, Formatter};
+
+use aws_lc_rs::aead::{
+    self, Aad, BoundKey, Nonce, NonceSequence, OpeningKey, SealingKey, UnboundKey, NONCE_LEN,
+};
+use aws_lc_rs::agreement;
+use aws_lc_rs::cipher::{AES_128_KEY_LEN, AES_256_KEY_LEN};
+use aws_lc_rs::digest::{SHA256_OUTPUT_LEN, SHA384_OUTPUT_LEN, SHA512_OUTPUT_LEN};
+use zeroize::Zeroize;
+
+use crate::crypto::aws_lc_rs::hmac::{HMAC_SHA256, HMAC_SHA384, HMAC_SHA512};
+use crate::crypto::hpke::{
+    EncapsulatedSecret, Hpke, HpkeOpener, HpkePrivateKey, HpkePublicKey, HpkeSealer, HpkeSuite,
+};
+use crate::crypto::tls13::{expand, HkdfExpander, HkdfPrkExtract, HkdfUsingHmac};
+use crate::msgs::enums::{HpkeAead, HpkeKdf, HpkeKem};
+use crate::msgs::handshake::HpkeSymmetricCipherSuite;
+use crate::{Error, OtherError};
+
+/// Default [RFC 9180] Hybrid Public Key Encryption (HPKE) suites supported by aws-lc-rs cryptography.
+pub static ALL_SUPPORTED_SUITES: &[&dyn Hpke] = &[
+    DH_KEM_P256_HKDF_SHA256_AES_128,
+    DH_KEM_P256_HKDF_SHA256_AES_256,
+    #[cfg(not(feature = "fips"))]
+    DH_KEM_P256_HKDF_SHA256_CHACHA20_POLY1305,
+    DH_KEM_P384_HKDF_SHA384_AES_128,
+    DH_KEM_P384_HKDF_SHA384_AES_256,
+    #[cfg(not(feature = "fips"))]
+    DH_KEM_P384_HKDF_SHA384_CHACHA20_POLY1305,
+    DH_KEM_P521_HKDF_SHA512_AES_128,
+    DH_KEM_P521_HKDF_SHA512_AES_256,
+    #[cfg(not(feature = "fips"))]
+    DH_KEM_P521_HKDF_SHA512_CHACHA20_POLY1305,
+    #[cfg(not(feature = "fips"))]
+    DH_KEM_X25519_HKDF_SHA256_AES_128,
+    #[cfg(not(feature = "fips"))]
+    DH_KEM_X25519_HKDF_SHA256_AES_256,
```

Comment:
Is this gated out because of X25519? That is inconsistent with what we're doing elsewhere. But having looked into it in detail this morning, I think what you're doing here is correct.

Reply:
> Is this gated out because of X25519? 

Correct.

>  But having looked into it in detail this morning, I think what you're doing here is correct.

I did some research at the time and arrived at the conclusion X25519 for kex wasn't permitted. I think only ed25519 for signatures is permitted. My gut instinct was to mark it as not FIPS approved and wait for someone with a stronger stomach for wading through NIST documents to make a case for why it is :laughing: 

This also has some interplay with the ECH spec, where [section 9](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-18#section-9) says:

> In the absence of an application profile standard specifying otherwise, a compliant ECH application MUST implement the following HPKE cipher suite:
>
> KEM: DHKEM(X25519, HKDF-SHA256) (see Section 7.1 of [HPKE])
> > KDF: HKDF-SHA256 (see Section 7.2 of [HPKE])
> AEAD: AES-128-GCM (see Section 7.3 of [HPKE])

I view "FIPS" as "an application profile standard specifying otherwise", and it felt like having ECH w/o X25519 for FIPS users was better than nothing. :man_shrugging: